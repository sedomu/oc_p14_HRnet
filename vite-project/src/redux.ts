import {configureStore, createSlice, type PayloadAction} from '@reduxjs/toolkit'
import {MOCKUP_DATA} from "@/config.ts";
import {employeesMockupData} from "@/data/employees_mockup.ts";

export type Employee = {
    firstName: string
    lastName: string
    dateOfBirth: string
    startDate: string
    street: string
    city: string
    state: string
    zipCode: number
    department: string
}

const initialState: Employee[] = []

if (MOCKUP_DATA){
    initialState.push(...employeesMockupData);
}

const employeesSlice = createSlice({
    name: 'employees',
    initialState: initialState,
    reducers: {
        addEmployee: (state, action: PayloadAction<Employee>) => {
            state.push(action.payload)
        }
    }
})

export const store = configureStore({
    reducer: {
        employees: employeesSlice.reducer
    }
})

export const { addEmployee } = employeesSlice.actions

export type RootState = ReturnType<typeof store.getState>;