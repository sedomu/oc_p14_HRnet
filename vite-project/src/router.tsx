import { createBrowserRouter } from "react-router-dom";
import Layout from "./layouts/Layout.jsx";
import Home from "@/pages/Home.tsx";
import { lazy, Suspense } from "react";

const EmployeesList = lazy(
    () => import("@/pages/EmployeesList.tsx")
);

const LazyEmployeesList = (
    <Suspense fallback={<div>Employees table is loading...</div>}>
        <EmployeesList />
    </Suspense>
);

export const router = createBrowserRouter([
    {
        path: "/",
        element: <Layout />,
        children: [
            { path: "/", element: <Home /> },
            {
                path: "/employee-list",
                element: LazyEmployeesList,
            },
        ],
    },
]);
