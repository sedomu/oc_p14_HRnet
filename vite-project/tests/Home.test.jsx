import { describe, it, expect, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import { Home } from "lucide-react";
import { RouterProvider } from "react-router-dom";
import { router } from "@/router.jsx";
import { Provider } from "react-redux";
import { store } from "@/redux.js";
import { userEvent } from "@testing-library/user-event";

vi.mock("@/components/form/Form.tsx", () => ({
    Form: ({ onSuccess }) => <button onClick={onSuccess}>Submit Form</button>,
}));

describe("Home", () => {
    it("should render correctly", () => {
        const { container } = render(
            <Provider store={store}>
                <RouterProvider router={router}>
                    <Home />
                </RouterProvider>
            </Provider>
        );

        expect(container).toMatchSnapshot();
    });
    it("should display the dialog box when form is successfully submitted", async () => {
        render(
            <Provider store={store}>
                <RouterProvider router={router}>
                    <Home />
                </RouterProvider>
            </Provider>
        );

        expect(() => screen.getByText("Employee Created!")).toThrow();

        await userEvent.click(screen.getByText("Submit Form"));

        expect(
            document.querySelector(".tiny-react-dialog__content > div")
                .textContent
        ).toBe("Employee Created!");
    });

    it("should close the dialog box when the user closes it (tested because of props drilling)", async () => {
        render(
            <Provider store={store}>
                <RouterProvider router={router}>
                    <Home />
                </RouterProvider>
            </Provider>
        );

        expect(() => screen.getByText("Employee Created!")).toThrow();

        await userEvent.click(screen.getByText("Submit Form"));

        expect(
            document.querySelector(".tiny-react-dialog__content > div")
                .textContent
        ).toBe("Employee Created!");

        await userEvent.click(
            document.querySelector(".tiny-react-dialog__close")
        );

        expect(() => screen.getByText("Employee Created!")).toThrow();
    });
});
