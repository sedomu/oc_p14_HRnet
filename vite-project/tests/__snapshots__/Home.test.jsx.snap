// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Home > should render correctly 1`] = `
<div>
  <div
    class="bg-muted flex min-h-svh flex-col items-center justify-center gap-6 p-6 md:p-10"
  >
    <a
      class="flex items-center gap-2 self-center font-medium"
      data-discover="true"
      href="/"
    >
      <img
        alt="logo"
        class="size-30"
        src="./logo.webp"
      />
    </a>
    <nav
      class="flex gap-4"
    >
      <button
        class="items-center justify-center whitespace-nowrap font-medium disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none shrink-0 [&_svg]:shrink-0 focus-visible:border-ring aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive border bg-background shadow-xs dark:bg-input/30 dark:border-input dark:hover:bg-input/50 h-9 has-[>svg]:px-3 data-[active=true]:focus:bg-accent data-[active=true]:hover:bg-accent data-[active=true]:bg-accent/50 data-[active=true]:text-accent-foreground hover:bg-accent hover:text-accent-foreground focus:bg-accent focus:text-accent-foreground focus-visible:ring-ring/50 [&_svg:not([class*='text-'])]:text-muted-foreground flex flex-col gap-1 rounded-sm p-2 text-sm transition-all outline-none focus-visible:ring-[3px] focus-visible:outline-1 [&_svg:not([class*='size-'])]:size-4"
        data-slot="button"
      >
        Home
      </button>
      <button
        class="items-center justify-center whitespace-nowrap font-medium disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none shrink-0 [&_svg]:shrink-0 focus-visible:border-ring aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive border bg-background shadow-xs dark:bg-input/30 dark:border-input dark:hover:bg-input/50 h-9 has-[>svg]:px-3 data-[active=true]:focus:bg-accent data-[active=true]:hover:bg-accent data-[active=true]:bg-accent/50 data-[active=true]:text-accent-foreground hover:bg-accent hover:text-accent-foreground focus:bg-accent focus:text-accent-foreground focus-visible:ring-ring/50 [&_svg:not([class*='text-'])]:text-muted-foreground flex flex-col gap-1 rounded-sm p-2 text-sm transition-all outline-none focus-visible:ring-[3px] focus-visible:outline-1 [&_svg:not([class*='size-'])]:size-4"
        data-slot="button"
      >
        Current Employees
      </button>
    </nav>
    <div
      class="flex w-full max-w-sm flex-col gap-6"
    >
      <button>
        Submit Form
      </button>
    </div>
  </div>
</div>
`;
